### PR TITLE
SE: Refactor CreateCfg method to get the SyntaxNode as input as is without having to retrieve its body.

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
Better review commit by commit.

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CFG/Roslyn/ControlFlowGraphCache.cs [hunk 1]
```diff
@@ -33,33 +33,40 @@ public abstract class ControlFlowGraphCacheBase
 
         public ControlFlowGraph FindOrCreate(SyntaxNode declaration, SemanticModel model, CancellationToken cancel)
         {
-            var rootSyntax = model.GetOperation(declaration).RootOperation().Syntax;
-            var nodeCache = compilationCache.GetValue(model.Compilation, x => new());
-            if (!nodeCache.TryGetValue(rootSyntax, out var wrapper))
+            if (model.GetOperation(declaration) is { } declarationOperation)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/Roslyn/ControlFlowGraphCache.cs:36]: Please do this and write some UT for it. We will need it for sonar-security.
```suggestion
var cfgNode = declaration switch 
{
    IndexerDeclarationSyntax indexer => indexer.ExpressionBody,
    PropertyDeclarationSyntax property => property.ExpressionBody,
    _ => declaration
};
            if (model.GetOperation(cfgNode) is { } declarationOperation)
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
We'll need one more functionality in this area

File: analyzers/src/SonarAnalyzer.CFG/Roslyn/ControlFlowGraphCache.cs (hunk 1 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/Roslyn/ControlFlowGraphCache.cs:36]: @pavel-mikula-sonarsource I'm doing this [in the runner here](https://github.com/SonarSource/sonar-dotnet/pull/8073/files#diff-7c2044a5e7097e4323205fc4ee570cc0d1bae3c95422ede523982ec15bd14ef4R89),

Just to be clear: the comment is to transfer the logic here right? Not to add it.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/Roslyn/ControlFlowGraphCache.cs:36]: Yes

File: analyzers/src/SonarAnalyzer.CFG/Roslyn/ControlFlowGraphCache.cs [hunk 2]
```diff
@@ -33,7 +34,14 @@ public abstract class ControlFlowGraphCacheBase
...
         {
-            if (model.GetOperation(declaration) is { } declarationOperation)
+            var cfgInputNode = declaration switch
+            {
+                IndexerDeclarationSyntax indexer => indexer.ExpressionBody,
+                PropertyDeclarationSyntax property => property.ExpressionBody,
+                _ => declaration
+            };
+
+            if (cfgInputNode is { } &&  model.GetOperation(cfgInputNode) is { } declarationOperation)
```
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/Roslyn/ControlFlowGraphCache.cs:44]: ExpressionBody can be null.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DeadStores.cs [hunk 3]
```diff
@@ -46,36 +46,27 @@ protected override void Initialize(SonarAnalysisContext context)
...
-            context.RegisterNodeAction(
-                c => CheckForDeadStores(c, c.SemanticModel.GetDeclaredSymbol(c.Node), ((LocalFunctionStatementSyntaxWrapper)c.Node).Body),
-                SyntaxKindEx.LocalFunctionStatement);
         }
 
-        private void CheckForDeadStores<T>(SonarSyntaxNodeReportingContext context, ISymbol symbol, Func<T, CSharpSyntaxNode> bodyOrExpressionBody) where T : SyntaxNode =>
-            CheckForDeadStores(context, symbol, bodyOrExpressionBody((T)context.Node));
-
-        private void CheckForDeadStores(SonarSyntaxNodeReportingContext context, ISymbol symbol, CSharpSyntaxNode node)
+        private void CheckForDeadStores(SonarSyntaxNodeReportingContext context, ISymbol symbol, SyntaxNode node)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DeadStores.cs:69]: It's already available in `context.Node`, no need to send it twice
```suggestion
        private void CheckForDeadStores(SonarSyntaxNodeReportingContext context, ISymbol symbol)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion.cs [hunk 4]
```diff
@@ -75,11 +77,11 @@ protected override void Initialize(SonarAnalysisContext context)
                 SyntaxKind.PropertyDeclaration);
         }
 
-        private void CheckForNoExitMethod(SonarSyntaxNodeReportingContext c, CSharpSyntaxNode body, SyntaxToken identifier)
+        private void CheckForNoExitMethod(SonarSyntaxNodeReportingContext c, CSharpSyntaxNode node, SyntaxToken identifier)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion.cs:80]: Duplicate
```suggestion
        private void CheckForNoExitMethod(SonarSyntaxNodeReportingContext c, SyntaxToken identifier)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion.cs [hunk 5]
```diff
@@ -75,11 +77,11 @@ protected override void Initialize(SonarAnalysisContext context)
...
         {
-            if (body != null && c.SemanticModel.GetDeclaredSymbol(c.Node) is IMethodSymbol symbol)
+            if (node != null && c.SemanticModel.GetDeclaredSymbol(c.Node) is IMethodSymbol symbol)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion.cs:82]: Impossible null
```suggestion
            if (c.SemanticModel.GetDeclaredSymbol(c.Node) is IMethodSymbol symbol)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs [hunk 6]
```diff
@@ -45,20 +45,20 @@ public sealed class MethodParameterUnused : MethodParameterUnusedBase
 
         protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterNodeAction(c =>
+            {
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:48]: This was shifted -4 and it shouldn't have been

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs [hunk 7]
```diff
@@ -126,8 +126,8 @@ private void ReportUnusedParametersOnMethod(MethodContext declaration)
 
         private void ReportOnDeadParametersAtEntry(MethodContext declaration, IImmutableList<IParameterSymbol> noReportOnParameters)
         {
-            var bodyNode = (CSharpSyntaxNode)declaration.Body ?? declaration.ExpressionBody;
-            if (bodyNode == null || declaration.Context.Node.IsKind(SyntaxKind.ConstructorDeclaration))
+            var declarationNode = (CSharpSyntaxNode)declaration.MethodDeclaration ?? declaration.LocalFunctionDeclaration;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:129]: Is the LocalFunctionDeclaration fixing old FNs? That would be nice to extract in a dedicated PR with issue

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs [hunk 8]
```diff
@@ -126,8 +126,8 @@ private void ReportUnusedParametersOnMethod(MethodContext declaration)
...
+            if (declarationNode == null || declaration.Context.Node.IsKind(SyntaxKind.ConstructorDeclaration))
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:130]: CaYC
```suggestion
            if (declarationNode is null || declaration.Context.Node.IsKind(SyntaxKind.ConstructorDeclaration))
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs (hunk 7 above)
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:129]: Let's get rid of the need to cast to `CSharpSyntaxNode` in a followup PR

Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:129]: For simplicity, there's just one node around here now
```suggestion
            var node = (CSharpSyntaxNode)declaration.MethodDeclaration ?? declaration.LocalFunctionDeclaration;
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs [hunk 9]
```diff
@@ -256,15 +256,23 @@ private class MethodContext
         {
             public readonly SonarSyntaxNodeReportingContext Context;
             public readonly IMethodSymbol Symbol;
+            public readonly BaseMethodDeclarationSyntax MethodDeclaration;
+            public readonly LocalFunctionStatementSyntaxWrapper LocalFunctionDeclaration;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:260]: Why are these split, when the only operation on them is to merge them?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs [hunk 10]
```diff
@@ -256,15 +256,23 @@ private class MethodContext
...
             public readonly ParameterListSyntax ParameterList;
             public readonly BlockSyntax Body;
             public readonly ArrowExpressionClauseSyntax ExpressionBody;
 
             public MethodContext(SonarSyntaxNodeReportingContext context, BaseMethodDeclarationSyntax declaration)
-                : this(context, declaration.ParameterList, declaration.Body, declaration.ExpressionBody()) { }
+                : this(context, declaration.ParameterList, declaration.Body, declaration.ExpressionBody())
+            {
+                MethodDeclaration = declaration;
+            }
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:269]: Should be arrow method. Same below

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantJumpStatement.cs [hunk 11]
```diff
@@ -35,27 +35,27 @@ public sealed class RedundantJumpStatement : SonarDiagnosticAnalyzer
         protected override void Initialize(SonarAnalysisContext context)
         {
             context.RegisterNodeAction(
-                c => CheckForRedundantJumps(c, ((BaseMethodDeclarationSyntax)c.Node).Body),
+                c => CheckForRedundantJumps(c, ((BaseMethodDeclarationSyntax)c.Node)),
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantJumpStatement.cs:38]: Useless cast, useless parameter
```suggestion
                CheckForRedundantJumps,
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs [hunk 12]
```diff
@@ -172,20 +172,68 @@ public void ArrowExpressionBody_WithNotExpectedNode_ReturnsNull()
...
+        }
+
+        [TestMethod]
+        public void CreateCfg_PropertyDeclartion_ReturnsCfg_CS()
+        {
+            const string code = """
+                public class Sample
+                {
+                    int field;
+                    public int Property => field + 42;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs:199]: To remove scaffolding noise
```suggestion
                    public int Property => 42;
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs [hunk 13]
```diff
@@ -172,20 +172,68 @@ public void ArrowExpressionBody_WithNotExpectedNode_ReturnsNull()
...
+            const string code = """
+                public class Sample
+                {
+                    public int Property {get; set;}
+                }
+                """;
+            var (tree, semanticModel) = TestHelper.CompileCS(code);
+            var node = tree.Single<SyntaxCS.PropertyDeclarationSyntax>();
+
+            ExtensionsCS.CreateCfg(node, semanticModel, default).Should().BeNull();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs:220]: This fragment seems duplicated many times. It could be `=> ValidateCreateCfg("""...` instead

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Some polishing

File: analyzers/src/SonarAnalyzer.CFG/Sonar/CSharpControlFlowGraph.cs [hunk 14]
```diff
@@ -18,18 +18,32 @@
...
+            var body = node switch
+            {
+                BaseMethodDeclarationSyntax n => (SyntaxNode)n.Body ?? n.ExpressionBody(),
+                PropertyDeclarationSyntax n => n.ExpressionBody?.Expression,
+                IndexerDeclarationSyntax n => n.ExpressionBody?.Expression,
+                AccessorDeclarationSyntax n => (SyntaxNode)n.Body ?? n.ExpressionBody(),
+                AnonymousFunctionExpressionSyntax n => n.Body,
+                ArrowExpressionClauseSyntax n => n,
+                _ when LocalFunctionStatementSyntaxWrapper.IsInstance(node) && (LocalFunctionStatementSyntaxWrapper)node is var local =>
+                    (SyntaxNode)local.Body ?? local.ExpressionBody,
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/Sonar/CSharpControlFlowGraph.cs:39]: It's not the prettiest line around, but it fits on a single line
```suggestion
                _ when LocalFunctionStatementSyntaxWrapper.IsInstance(node) && (LocalFunctionStatementSyntaxWrapper)node is var local => (SyntaxNode)local.Body ?? local.ExpressionBody,
```

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 15]
```diff
@@ -72,27 +72,28 @@ protected static bool IsEnabled(SonarSyntaxNodeReportingContext context, Diagnos
...
-            Analyze(analysisContext, context, body, symbol);
+            Analyze(analysisContext, context, symbol);
         }
     }
 
-    protected void Analyze(SonarAnalysisContext analysisContext, SonarSyntaxNodeReportingContext nodeContext, SyntaxNode body, ISymbol symbol)
+    protected void Analyze(SonarAnalysisContext analysisContext, SonarSyntaxNodeReportingContext nodeContext, ISymbol symbol)
     {
-        if (body is { ContainsDiagnostics: false })
+        var node = nodeContext.Node;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:85]: This is a weird (useless) rename.
```suggestion
```

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 16]
```diff
@@ -72,27 +72,28 @@ protected static bool IsEnabled(SonarSyntaxNodeReportingContext context, Diagnos
...
+        if (node is { ContainsDiagnostics: false })
         {
-            AnalyzeSonar(nodeContext, body, symbol);
+            AnalyzeSonar(nodeContext, node, symbol);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:88]: This looks useless now
```suggestion
            AnalyzeSonar(nodeContext, symbol);
```

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 17]
```diff
@@ -72,27 +72,28 @@ protected static bool IsEnabled(SonarSyntaxNodeReportingContext context, Diagnos
...
             if (ControlFlowGraph.IsAvailable)
             {
-                AnalyzeRoslyn(analysisContext, nodeContext, body, symbol);
+                AnalyzeRoslyn(analysisContext, nodeContext, node, symbol);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:91]: Same duplicate here
```suggestion
                AnalyzeRoslyn(analysisContext, nodeContext, symbol);
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs (hunk 9 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs]: Indeed - refactored.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs (hunk 10 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs]: code removed - comment not relevant anymore.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DeadStores.cs [hunk 18]
```diff
@@ -46,37 +46,29 @@ protected override void Initialize(SonarAnalysisContext context)
...
-                SyntaxKindEx.LocalFunctionStatement);
         }
 
-        private void CheckForDeadStores<T>(SonarSyntaxNodeReportingContext context, ISymbol symbol, Func<T, CSharpSyntaxNode> bodyOrExpressionBody) where T : SyntaxNode =>
-            CheckForDeadStores(context, symbol, bodyOrExpressionBody((T)context.Node));
-
-        private void CheckForDeadStores(SonarSyntaxNodeReportingContext context, ISymbol symbol, CSharpSyntaxNode node)
+        private void CheckForDeadStores(SonarSyntaxNodeReportingContext context, ISymbol symbol)
         {
+            var node = context.Node;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DeadStores.cs:71]: I would rather replace the usages with `context.Node`
```suggestion
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantJumpStatement.cs [hunk 19]
```diff
@@ -35,34 +35,35 @@ public sealed class RedundantJumpStatement : SonarDiagnosticAnalyzer
...
+                CheckForRedundantJumps,
                 SyntaxKind.AnonymousMethodExpression,
                 SyntaxKind.SimpleLambdaExpression,
                 SyntaxKind.ParenthesizedLambdaExpression);
         }
 
-        private static void CheckForRedundantJumps(SonarSyntaxNodeReportingContext context, CSharpSyntaxNode node)
+        private static void CheckForRedundantJumps(SonarSyntaxNodeReportingContext context)
         {
+            var node = context.Node;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantJumpStatement.cs:66]: Use the context directly
```suggestion
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Intermediate round after review 2, batch 2

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs (hunk 13 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs:220]: Not so trivial - the syntax type based on which we get the node for the cfg creation changes + the extensions; ' namespace  (depending if you want a cfg for CS or VB code).

I'll be happy to do this extraction in another PR.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs (hunk 7 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs]: No, I found my answer later. So the resolving was intentional.

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs [hunk 20]
```diff
@@ -1013,7 +1061,7 @@ public void ParameterList_PrimaryConstructors(string type)
             var node = NodeBetweenMarkers($$"""
                 $$public {{type}} C(int p)
                 {
-                    
+
```
Comment (mary-georgiou-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs:1064]: What is this notation?
`$$int i;$$`

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs (hunk 7 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs]: Added a task to my calendar for this.
Not sure when I'll tackle it really :/

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs [hunk 21]
```diff
@@ -126,8 +127,8 @@ private void ReportUnusedParametersOnMethod(MethodContext declaration)
 
         private void ReportOnDeadParametersAtEntry(MethodContext declaration, IImmutableList<IParameterSymbol> noReportOnParameters)
         {
-            var bodyNode = (CSharpSyntaxNode)declaration.Body ?? declaration.ExpressionBody;
-            if (bodyNode == null || declaration.Context.Node.IsKind(SyntaxKind.ConstructorDeclaration))
+            var node = (CSharpSyntaxNode)(declaration.Body?.Parent ?? declaration.ExpressionBody.Parent);
+            if (node is null || declaration.Context.Node.IsKind(SyntaxKind.ConstructorDeclaration))
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:131]: I don't like the parent thing, because it's like "rovnák na ohejbák" (something like a streightener used to undo work done by a bender)

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs [hunk 22]
```diff
@@ -172,20 +172,67 @@ public void ArrowExpressionBody_WithNotExpectedNode_ReturnsNull()
...
         {
-            const string code = @"
-public class Sample
-{
-    public void Main()
-    {
-        var x = 42;
-    }
-}";
+            var code = """
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs:177]: Also below

File: analyzers/src/SonarAnalyzer.CFG/Sonar/CSharpControlFlowGraph.cs [hunk 23]
```diff
@@ -18,18 +18,31 @@
...
             cfg = null;
+            var body = node switch
+            {
+                BaseMethodDeclarationSyntax n => (SyntaxNode)n.Body ?? n.ExpressionBody(),
+                PropertyDeclarationSyntax n => n.ExpressionBody?.Expression,
+                IndexerDeclarationSyntax n => n.ExpressionBody?.Expression,
+                AccessorDeclarationSyntax n => (SyntaxNode)n.Body ?? n.ExpressionBody(),
+                AnonymousFunctionExpressionSyntax n => n.Body,
+                ArrowExpressionClauseSyntax n => n,
+                _ when LocalFunctionStatementSyntaxWrapper.IsInstance(node) && (LocalFunctionStatementSyntaxWrapper)node is var local => (SyntaxNode)local.Body ?? local.ExpressionBody,
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/Sonar/CSharpControlFlowGraph.cs:38]: `.IsKind` is much cheaper than `IsInstance`

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 24]
```diff
@@ -72,47 +72,47 @@ protected static bool IsEnabled(SonarSyntaxNodeReportingContext context, Diagnos
...
             if (ControlFlowGraph.IsAvailable)
             {
-                AnalyzeRoslyn(analysisContext, nodeContext, body, symbol);
+                AnalyzeRoslyn(analysisContext, context, symbol);
             }
         }
     }
 
-    private void AnalyzeRoslyn(SonarAnalysisContext analysisContext, SonarSyntaxNodeReportingContext nodeContext, SyntaxNode body, ISymbol symbol)
+    private void AnalyzeRoslyn(SonarAnalysisContext analysisContext, SonarSyntaxNodeReportingContext context, ISymbol symbol)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:95]: The `node` part of `nodeContext` name was important here. Because there's also `analysisContext` and we need to clearly distinguish those two apart.

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 25]
```diff
@@ -72,47 +72,47 @@ protected static bool IsEnabled(SonarSyntaxNodeReportingContext context, Diagnos
...
-        if (getBody((TNode)context.Node) is { } body && context.SemanticModel.GetDeclaredSymbol(context.Node) is { } symbol)
+        if (context.SemanticModel.GetDeclaredSymbol(context.Node) is { } symbol)
         {
-            Analyze(analysisContext, context, body, symbol);
+            Analyze(analysisContext, context, symbol);
         }
     }
 
-    protected void Analyze(SonarAnalysisContext analysisContext, SonarSyntaxNodeReportingContext nodeContext, SyntaxNode body, ISymbol symbol)
+    protected void Analyze(SonarAnalysisContext analysisContext, SonarSyntaxNodeReportingContext context, ISymbol symbol)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:83]: The `node` part of `nodeContext` name was important here. Because there's also `analysisContext` and we need to clearly distinguish those two apart.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
LIttle more polishing

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs (hunk 20 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs:1063]: See `NodeBetweenMarkers`. `$$` is used as a (somewhat confusing) UT scaffolding annotation

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs [hunk 26]
```diff
@@ -260,13 +261,7 @@ private class MethodContext
...
             public readonly ArrowExpressionClauseSyntax ExpressionBody;
 
-            public MethodContext(SonarSyntaxNodeReportingContext context, BaseMethodDeclarationSyntax declaration)
-                : this(context, declaration.ParameterList, declaration.Body, declaration.ExpressionBody()) { }
-
-            public MethodContext(SonarSyntaxNodeReportingContext context, LocalFunctionStatementSyntaxWrapper declaration)
-                : this(context, declaration.ParameterList, declaration.Body, declaration.ExpressionBody) { }
-
-            private MethodContext(SonarSyntaxNodeReportingContext context, ParameterListSyntax parameterList, BlockSyntax body, ArrowExpressionClauseSyntax expressionBody)
+            public MethodContext(SonarSyntaxNodeReportingContext context, ParameterListSyntax parameterList, BlockSyntax body, ArrowExpressionClauseSyntax expressionBody)
```
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:264]: re-added

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs (hunk 13 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs:219]: This PR starts being big and long-lived - I"ll follow up with another PR. (added a task to myself)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs (hunk 21 above)
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:131]: I don't like it either - but what I did before in order to get rid of it - was a big bad indirection.
I don't see an easy way to get rid of this that doesn't require creating a new PR practically from this PR from my side.
This is a hardening ticket IMO - unless you see something I don't.

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs (hunk 22 above)
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs:177]: I found it noisy - and not sure how much it adds here to have the strings set on compile time.
Not strongly set on this though - I reverted.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs [hunk 27]
```diff
@@ -69,25 +69,26 @@ private static MethodContext CreateContext(SonarSyntaxNodeReportingContext c)
             {
                 return new MethodContext(c, method);
             }
-            else if (c.Node.Kind() == SyntaxKindEx.LocalFunctionStatement)
+            else if (c.Node.IsKind(SyntaxKindEx.LocalFunctionStatement))
             {
-                return new MethodContext(c, (LocalFunctionStatementSyntaxWrapper)c.Node);
+                var localFunction = (LocalFunctionStatementSyntaxWrapper)c.Node;
+                return new MethodContext(c, localFunction);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:75]: You can fully revert the line. The rename isn't needed
```suggestion
                return new MethodContext(c, (LocalFunctionStatementSyntaxWrapper)c.Node);
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs [hunk 28]
```diff
@@ -252,21 +249,20 @@ private static bool IsStandaloneExpression(ExpressionSyntax expression)
...
-        private class MethodContext
+        private sealed class MethodContext
         {
             public readonly SonarSyntaxNodeReportingContext Context;
             public readonly IMethodSymbol Symbol;
             public readonly ParameterListSyntax ParameterList;
             public readonly BlockSyntax Body;
             public readonly ArrowExpressionClauseSyntax ExpressionBody;
-
             public MethodContext(SonarSyntaxNodeReportingContext context, BaseMethodDeclarationSyntax declaration)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs:259]: There's missing EOL before this line that was removed in this PR

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs (hunk 7 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs]: That was done in this PR in the end. It was easier than expected.

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs (hunk 13 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs:219]: https://github.com/SonarSource/sonar-dotnet/pull/8355

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs (hunk 7 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs]: #8374 is what I meant. I originally thought it's much bigger, but the core effort was already done in the past.

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs]: Right.
I thought the comment was dedicated to `MethodParameterUnused`

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MethodParameterUnused.cs]: Yep, I wasn't clear enough with my original comment.
